racket/collects/racket/private/prop-template-metafunction.rkt

```diff
@@ -0,0 +1,88 @@
+(module prop-template-metafunction '#%kernel
+  (#%provide (protect template-metafunction
+                      prop:template-metafunction
+                      (rename prop:template-metafunction? template-metafunction?)
+                      template-metafunction-accessor))
+  
+  ;; The prop:template-metafunction structure type property can contain one of:
+  ;;  - an identifier bound to the run-time metafunction procedure
+  ;;  - the index of a field containing such an identifier
+  ;;  - a function that takes the structure instance and produces such an
+  ;;    identifier
+  ;; At run-time, when processing the template, the syntax object whose first
+  ;; element is a metafunction identifier is passed to this metafunction
+  ;; procedure.
+
+  ;; Internally, the value inside the property will be stored as a function
+  ;; that takes the structure instance and produces the identifier,
+  ;; so the first two cases of that `one-of` will be "normalized" to the
+  ;; function case by `prop:template-metafunction-guard`.
+
+  (define-values (prop:template-metafunction-guard)
+    (lambda (val struct-type-info-list)
+      (if (exact-nonnegative-integer? val)
+          ;; Field index, the field must contain an identifier bound to the run-time
+          ;; metafunction procedure.
+          (let-values ([(make-struct-accessor) (cadddr struct-type-info-list)])
+            (let-values ([(accessor) (make-struct-field-accessor make-struct-accessor val)])
+              (λ (instance)
+                (let-values ([(metafunction-internal-id) (accessor instance)])
+                  ;; Check that the value contained in the field is correct
+                  ;; (if (identifier? metafunction-internal-id) ...)
+                  (if (if (syntax? metafunction-internal-id)
+                          (symbol? (syntax-e metafunction-internal-id))
+                          #f)
+                      metafunction-internal-id
+                      (raise-argument-error
```

Comment:
`raise-argument-error` should be used with a contract, not a statement. Maybe use `raise-arguments-error` instead (note plural).